File: typescript/src/slo.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * SLO-oriented telemetry helpers (RED/USE baseline).
 * Mirrors Python provide.telemetry.slo.
 */

import {
  type CounterInstrument,
  type GaugeInstrument,
  type HistogramInstrument,
  counter,
  gauge,
  histogram,
} from './metrics';

const _counters = new Map<string, CounterInstrument>();
const _histograms = new Map<string, HistogramInstrument>();
const _gauges = new Map<string, GaugeInstrument>();

function _lazyCounter(name: string, description: string): CounterInstrument {
  let c = _counters.get(name);
  if (!c) {
    c = counter(name, { description });
    _counters.set(name, c);
  }
  return c;
}

function _lazyHistogram(name: string, description: string, unit: string): HistogramInstrument {
  let h = _histograms.get(name);
  if (!h) {
    h = histogram(name, { description, unit });
    _histograms.set(name, h);
  }
  return h;
}

function _lazyGauge(name: string, description: string, unit: string): GaugeInstrument {
  let g = _gauges.get(name);
  if (!g) {
    g = gauge(name, { description, unit });
    _gauges.set(name, g);
  }
  return g;
}

export function recordRedMetrics(opts: {
  route: string;
  method: string;
  statusCode: number;
  durationMs: number;
}): void {
  if (!getConfig().sloEnableRedMetrics) return;
  const attrs = {
    route: opts.route,
    method: opts.method,
    status_code: String(opts.statusCode),
  };
  _lazyCounter('http.requests.total', 'Total HTTP requests').add(1, attrs);
  if (opts.statusCode >= 500) {
    // Stryker disable next-line StringLiteral: error counter description is not tested
    _lazyCounter('http.errors.total', 'Total HTTP errors').add(1, attrs);
  }
  _lazyHistogram('http.request.duration_ms', 'HTTP request latency', 'ms').record(
    opts.durationMs,
    attrs,
  );
}

export function recordUseMetrics(opts: {
  resource: string;
  utilization: number;
  unit?: string;
}): void {
  const g = _lazyGauge('resource.utilization', 'Resource utilization', opts.unit ?? '%');
  g.set(opts.utilization, { resource: opts.resource });
}

export interface ErrorClassification {
  errorType: 'server' | 'client' | 'timeout' | 'unknown';
  errorCode: number;
  errorName: string;
  category: 'server_error' | 'client_error' | 'timeout' | 'unknown';
  severity: 'critical' | 'warning' | 'info' | 'unknown';
  // OTel-aligned keys for cross-language parity with Go/Python
  'error.type': string;
  'error.category': string;
  'error.severity': string;
  'http.status_code': string;
}

export function classifyError(excName: string, statusCode: number): ErrorClassification {
  const isTimeout = statusCode === 0 || excName.toLowerCase().includes('timeout');

  if (isTimeout) {
    return {
      errorType: 'timeout',
      errorCode: statusCode,
      errorName: excName,
      category: 'timeout',
      severity: 'info',
      'error.type': excName,
      'error.category': 'timeout',
      'error.severity': 'info',
      'http.status_code': String(statusCode),
    };
  }
  if (statusCode >= 500) {
    return {
      errorType: 'server',
      errorCode: statusCode,
      errorName: excName,
      category: 'server_error',
      severity: 'critical',
      'error.type': excName,
      'error.category': 'server_error',
      'error.severity': 'critical',
      'http.status_code': String(statusCode),
    };
  }
  if (statusCode >= 400) {
    const sev = statusCode === 429 ? 'critical' : 'warning';
    return {
      errorType: 'client',
      errorCode: statusCode,
      errorName: excName,
      category: 'client_error',
      severity: sev as 'critical' | 'warning',
      'error.type': excName,
      'error.category': 'client_error',
      'error.severity': sev,
      'http.status_code': String(statusCode),
    };
  }
  return {
    errorType: 'unknown',
    errorCode: statusCode,
    errorName: excName,
    category: 'unknown',
    severity: 'unknown',
    'error.type': excName,
    'error.category': 'unknown',
    'error.severity': 'unknown',
    'http.status_code': String(statusCode),
  };
}

export function _resetSloForTests(): void {
  _counters.clear();
  _histograms.clear();
  _gauges.clear();
}

File: go/slo.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

package telemetry

import (
	"context"
	"log/slog"
	"strconv"
	"strings"
)

// Error classification category constants.
const (
	_errCatClientError  = "client_error"
	_errCatServerError  = "server_error"
	_errCatTimeout      = "timeout"
	_errCatUnknown      = "unknown"
	_errSevCritical     = "critical"
	_errSevWarning      = "warning"
	_errSevInfo         = "info"
)

// Package-level RED metric instruments (Rate, Errors, Duration).
var (
	_redRequestCounter    Counter   = NewCounter("slo.red.requests", WithDescription("RED request rate"))
	_redErrorCounter      Counter   = NewCounter("slo.red.errors", WithDescription("RED error count"))
	_redDurationHistogram Histogram = NewHistogram("slo.red.duration_ms", WithDescription("RED request duration"), WithUnit("ms"))
)

// Package-level USE metric instrument (Utilization, Saturation, Errors).
var _useUtilizationGauge Gauge = NewGauge("slo.use.utilization", WithDescription("USE utilization percentage"), WithUnit("%"))

// ClassifyError returns a map of error taxonomy fields for the given exception
// name and HTTP status code.
//
// Keys returned: "error.type", "http.status_code", "error.category", "error.severity".
func ClassifyError(excName string, statusCode int) map[string]string {
	result := map[string]string{
		"error.type":       excName,
		"http.status_code": strconv.Itoa(statusCode),
	}

	isTimeout := statusCode == 0 || strings.Contains(strings.ToLower(excName), "timeout")

	switch {
	case isTimeout:
		result["error.category"] = _errCatTimeout
		result["error.severity"] = _errSevInfo
	case statusCode >= 500:
		result["error.category"] = _errCatServerError
		result["error.severity"] = _errSevCritical
	case statusCode >= 400:
		result["error.category"] = _errCatClientError
		if statusCode == 429 {
			result["error.severity"] = _errSevCritical
		} else {
			result["error.severity"] = _errSevWarning
		}
	default:
		result["error.category"] = _errCatUnknown
		result["error.severity"] = _errSevInfo
	}

	return result
}

// RecordREDMetrics records Rate/Errors/Duration metrics for an HTTP request.
// It increments the request counter, conditionally increments the error counter
// for status codes >= 400, and records the duration.
func RecordREDMetrics(route, method string, statusCode int, durationMs float64) {
	ctx := context.Background()
	attrs := []slog.Attr{
		slog.String("route", route),
		slog.String("method", method),
		slog.String("status_code", strconv.Itoa(statusCode)),
	}
	_redRequestCounter.Add(ctx, 1, attrs...)
	if statusCode >= 400 {
		_redErrorCounter.Add(ctx, 1, attrs...)
	}
	_redDurationHistogram.Record(ctx, durationMs, attrs...)
}

// RecordUSEMetrics records the Utilization metric for a named resource.
func RecordUSEMetrics(resource string, utilizationPercent int) {
	ctx := context.Background()
	_useUtilizationGauge.Set(ctx, float64(utilizationPercent), slog.String("resource", resource))
}
